File: lab2/gr9sub2/МасюкВС/src/fork_example.py
import os
import sys
import time


def print_process_info(name):
    print(f"{name}: PID = {os.getpid()}, PPID = {os.getppid()}", flush=True)
    sys.exit(0)


def execute_parent_and_children():
    print(f"Parent: My PID is {os.getpid()}", flush=True)
    first_child_pid = -1
    second_child_pid = -1

    first_child_pid = os.fork()

    if first_child_pid == 0:
        print_process_info("Child A")

    if first_child_pid > 0:
        second_child_pid = os.fork()
        if second_child_pid == 0:
            print_process_info("Child B")

    if first_child_pid > 0 and second_child_pid > 0:
        child_1_pid, child_1_status = os.waitpid(first_child_pid, 0)
        child_1_exit_code = os.WEXITSTATUS(child_1_status)
        print(f"Parent: Child A (PID={child_1_pid}) finished with status {child_1_exit_code}", flush=True)

        child_2_pid, child_2_status = os.waitpid(second_child_pid, 0)
        child_2_exit_code = os.WEXITSTATUS(child_2_status)
        print(f"Parent: Child B (PID={child_2_pid}) finished with status {child_2_exit_code}", flush=True)

        print("Parent: All children have finished", flush=True)


if __name__ == "__main__":
    execute_parent_and_children()

File: lab2/gr8sub1/Дмитриева_Полина/src/fork_example.cpp
#include <iostream>
#include <unistd.h>
#include <sys/wait.h>
#include <cstdlib>

int main(){
std::cout<<"Parent pro# cess started (PID="<<getpid()<<")"<<std::endl;

pid_t child_a = fork();

if(child_a==-1){
std::cerr<<"First fork failed!"<<std::endl;
exit(1);
}
else if(child_a==0){
std::cout<<"Child_A:My PID is "<<getpid()<<",My Parent`s PID is "<<getppid()<<std::endl;
exit(0);
}

pid_t child_b=fork();

if(child_b==-1){
std::cerr<<"Second fork failed!"<<std::endl;
exit(1);
}
else if(child_b==0){
std::cout<<"Child_B: My PID is "<<getpid()<<",My Parent`s PID is "<<getppid()<<std::endl;
exit(0);
}
int status;
waitpid(child_a,&status,0);
std::cout<<"Parent: Child A(PID="<<child_a<<") finished with status: "<<WEXITSTATUS(status)<<std::endl;

waitpid(child_b,&status,0);
std::cout<<"Parent: Child B (PID="<<child_b<<") finished with status: "<<WEXITSTATUS(status)<<std::endl;

std::cout<<"Parent: All children finished. Exiting."<<std::endl;

return 0;
}
